File: samples/language-impl-comparison/elixir/algebraic_effects.exs
# 代数的効果を使うミニ言語 - Elixir 版
# Reml との比較: プロセスベースの効果エミュレーション

defmodule AlgebraicEffects do
  @moduledoc """
  代数的効果を Elixir のプロセスメッセージングでエミュレートする実装。

  Reml の代数的効果システムとの対比：
  - Reml: 言語レベルで effect/handler をサポート
  - Elixir: プロセス + GenServer でステートマシンとして実装
  - 効果は {:effect, :name, args} メッセージとして表現
  """

  # ミニ言語の式定義
  defmodule Expr do
    defstruct [:type, :args]

    def lit(n), do: %Expr{type: :lit, args: [n]}
    def var(name), do: %Expr{type: :var, args: [name]}
    def add(left, right), do: %Expr{type: :add, args: [left, right]}
    def mul(left, right), do: %Expr{type: :mul, args: [left, right]}
    def divide(left, right), do: %Expr{type: :div, args: [left, right]}
    def get(), do: %Expr{type: :get, args: []}
    def put(expr), do: %Expr{type: :put, args: [expr]}
    def fail(msg), do: %Expr{type: :fail, args: [msg]}
    def choose(left, right), do: %Expr{type: :choose, args: [left, right]}
  end

  # 効果ハンドラーのプロトコル
  defmodule EffectHandler do
    @doc """
    効果を処理する GenServer ベースのハンドラー。

    State: 可変状態を管理
    Except: 例外を捕捉して Result に変換
    Choose: 非決定的選択をリストで収集
    """

    use GenServer

    # クライアント API

    def start_link(init_state) do
      GenServer.start_link(__MODULE__, %{state: init_state, results: []})
    end

    def get_state(pid) do
      GenServer.call(pid, :get)
    end

    def put_state(pid, new_state) do
      GenServer.call(pid, {:put, new_state})
    end

    def raise_error(pid, msg) do
      GenServer.call(pid, {:raise, msg})
    end

    def choose(pid, left, right) do
      GenServer.call(pid, {:choose, left, right})
    end

    def get_final_state(pid) do
      GenServer.call(pid, :final_state)
    end

    # サーバーコールバック

    @impl true
    def init(initial_state) do
      {:ok, initial_state}
    end

    @impl true
    def handle_call(:get, _from, state) do
      {:reply, {:ok, state.state}, state}
    end

    @impl true
    def handle_call({:put, new_state}, _from, state) do
      {:reply, :ok, %{state | state: new_state}}
    end

    @impl true
    def handle_call({:raise, msg}, _from, state) do
      {:reply, {:error, msg}, state}
    end

    @impl true
    def handle_call({:choose, left, right}, _from, state) do
      # 非決定的選択: 両方の結果を収集
      {:reply, {:choices, [left, right]}, state}
    end

    @impl true
    def handle_call(:final_state, _from, state) do
      {:reply, state, state}
    end
  end

  # 式の評価関数（効果を持つ）
  @doc """
  式を評価する。効果は handler プロセス経由で処理される。

  Reml との違い:
  - Reml: perform Except.raise(msg) で効果を送出
  - Elixir: GenServer.call(handler, {:raise, msg}) でメッセージング
  """
  def eval(expr, env, handler) do
    case expr.type do
      :lit ->
        [n] = expr.args
        {:ok, n}

      :var ->
        [name] = expr.args
        case Enum.find(env, fn {k, _v} -> k == name end) do
          {_k, v} -> {:ok, v}
          nil -> EffectHandler.raise_error(handler, "未定義変数: #{name}")
        end

      :add ->
        [left, right] = expr.args
        with {:ok, l} <- eval(left, env, handler),
             {:ok, r} <- eval(right, env, handler) do
          {:ok, l + r}
        end

      :mul ->
        [left, right] = expr.args
        with {:ok, l} <- eval(left, env, handler),
             {:ok, r} <- eval(right, env, handler) do
          {:ok, l * r}
        end

      :div ->
        [left, right] = expr.args
        with {:ok, l} <- eval(left, env, handler),
             {:ok, r} <- eval(right, env, handler) do
          if r == 0 do
            EffectHandler.raise_error(handler, "ゼロ除算")
          else
            {:ok, div(l, r)}
          end
        end

      :get ->
        EffectHandler.get_state(handler)

      :put ->
        [e] = expr.args
        with {:ok, v} <- eval(e, env, handler),
             :ok <- EffectHandler.put_state(handler, v) do
          {:ok, v}
        end

      :fail ->
        [msg] = expr.args
        EffectHandler.raise_error(handler, msg)

      :choose ->
        [left, right] = expr.args
        with {:ok, l} <- eval(left, env, handler),
             {:ok, r} <- eval(right, env, handler) do
          {:choices, [l, r]}
        end
    end
  end

  # すべての効果をハンドル（State + Except + Choose）
  @doc """
  すべての効果を処理して結果を返す。

  Reml の handle ... do ... do ... に相当するが、
  Elixir ではプロセス起動 → 評価 → 結果収集の手順が必要。
  """
  def run_with_all_effects(expr, env, init_state) do
    {:ok, handler} = EffectHandler.start_link(init_state)

    result = case eval(expr, env, handler) do
      {:ok, value} ->
        final = EffectHandler.get_final_state(handler)
        {:ok, [{value, final.state}]}

      {:error, msg} ->
        {:error, msg}

      {:choices, values} ->
        final = EffectHandler.get_final_state(handler)
        results = Enum.map(values, fn v -> {v, final.state} end)
        {:ok, results}
    end

    GenServer.stop(handler)
    result
  end

  # テストケース
  def example_expressions do
    [
      {"単純な加算", Expr.add(Expr.lit(10), Expr.lit(20))},
      {"乗算と除算", Expr.divide(Expr.mul(Expr.lit(6), Expr.lit(7)), Expr.lit(2))},
      {"状態の取得", Expr.add(Expr.get(), Expr.lit(5))},
      {"状態の更新", Expr.put(Expr.add(Expr.get(), Expr.lit(1)))},
      {"ゼロ除算エラー", Expr.divide(Expr.lit(10), Expr.lit(0))},
      {"非決定的選択", Expr.choose(Expr.lit(1), Expr.lit(2))},
      {"複雑な例", Expr.add(
        Expr.choose(Expr.lit(10), Expr.lit(20)),
        Expr.put(Expr.add(Expr.get(), Expr.lit(1)))
      )}
    ]
  end

  # テスト実行
  def run_examples do
    examples = example_expressions()
    env = []
    init_state = 0

    Enum.each(examples, fn {name, expr} ->
      IO.puts("--- #{name} ---")
      case run_with_all_effects(expr, env, init_state) do
        {:ok, results} ->
          Enum.each(results, fn {value, state} ->
            IO.puts("  結果: #{value}, 状態: #{state}")
          end)

        {:error, err} ->
          IO.puts("  エラー: #{err}")
      end
    end)
  end
end

# Reml との比較メモ:
#
# 1. **効果の表現**
#    Reml: 言語レベルの effect/handler 構文
#    Elixir: プロセスメッセージング（GenServer）でエミュレート
#    - Reml はコンパイラが効果を追跡・最適化
#    - Elixir は実行時にプロセス間通信が発生（オーバーヘッド大）
#
# 2. **ハンドラーの合成**
#    Reml: handle state_handler(init) do ... で宣言的
#    Elixir: プロセス起動 → 評価 → 停止の手続き的フロー
#    - Reml の方がシンプルで理解しやすい
#
# 3. **非決定性の扱い**
#    Reml: choose_handler でリストを自動生成
#    Elixir: {:choices, [left, right]} を手動で管理
#    - Reml はハンドラーが自動で分岐を収集
#    - Elixir は明示的なパターンマッチが必要
#
# 4. **型安全性**
#    Reml: 効果が型レベルで追跡される
#    Elixir: 動的型付けのため実行時エラーのリスク
#
# 5. **パフォーマンス**
#    Reml: 効果はコンパイル時に最適化可能
#    Elixir: プロセス通信のオーバーヘッドが常に発生
#
# **結論**:
# Elixir は並行性に強いが、代数的効果の表現力では Reml に劣る。
# プロセスモデルは効果のエミュレーションに使えるが、
# 型安全性・パフォーマンス・可読性の面で Reml の方が優れている。

# テスト実行例
# AlgebraicEffects.run_examples()

File: examples/language-impl-comparison/haskell/algebraic_effects.hs
-- 代数的効果を使うミニ言語 - Haskell 版
-- Reml との比較: モナド変換子による効果のエミュレーション

{-# LANGUAGE DeriveFunctor #-}

module AlgebraicEffects where

import Control.Monad (foldM)
import Control.Monad.State (StateT, evalStateT, get, modify, put)
import Control.Monad.Except (ExceptT, runExceptT, throwError)
import Control.Monad.Trans.Class (lift)

-- ミニ言語の式定義
data Expr
  = Lit Int
  | Var String
  | Add Expr Expr
  | Mul Expr Expr
  | Div Expr Expr
  | Get
  | Put Expr
  | Fail String
  | Choose Expr Expr
  deriving (Show)

type Env = [(String, Int)]

-- 効果をモナド変換子スタックで表現
-- State<Int> × Except<String> × Choose（リスト）
-- Reml: with State<Int>, Except<String>, Choose
-- Haskell: StateT Int (ExceptT String []) a
type EffectM a = StateT Int (ExceptT String []) a

-- 式の評価関数（効果を持つ）
--
-- Reml の perform に相当する操作をモナド操作で記述：
-- - get/put: StateT の操作
-- - throwError: ExceptT の操作
-- - lift . lift: リストモナド（非決定性）への持ち上げ
eval :: Expr -> Env -> EffectM Int
eval (Lit n) _ = return n

eval (Var name) env =
  case lookup name env of
    Just value -> return value
    Nothing -> throwError $ "未定義変数: " ++ name

eval (Add left right) env = do
  l <- eval left env
  r <- eval right env
  return (l + r)

eval (Mul left right) env = do
  l <- eval left env
  r <- eval right env
  return (l * r)

eval (Div left right) env = do
  l <- eval left env
  r <- eval right env
  if r == 0
    then throwError "ゼロ除算"
    else return (l `div` r)

eval Get _ = get

eval (Put e) env = do
  v <- eval e env
  put v
  return v

eval (Fail msg) _ = throwError msg

eval (Choose left right) env = do
  -- 非決定的選択: lift を2回使ってリストモナドに到達
  -- Reml の choose_handler に相当する部分が複雑
  st <- get
  let leftM = evalStateT (eval left env) st
  let rightM = evalStateT (eval right env) st
  results <- lift $ lift [leftM, rightM]
  case results of
    Left err -> throwError err
    Right (value, newState) -> do
      put newState
      return value

-- すべての効果を処理して結果を返す
--
-- Reml の handle ... do ... do ... に相当するが、
-- Haskell ではモナド変換子のスタックを順に実行。
runWithAllEffects :: Expr -> Env -> Int -> Either String [(Int, Int)]
runWithAllEffects expr env initState =
  case runExceptT $ evalStateT (eval expr env) initState of
    [] -> Left "選択肢なし"
    results -> sequence results >>= \vals -> Right [(v, s) | (v, s) <- vals]

-- より簡易版: State のみをハンドル
runWithState :: Expr -> Env -> Int -> ExceptT String [] (Int, Int)
runWithState expr env initState = do
  value <- evalStateT (eval expr env) initState
  return (value, initState)

-- テストケース
exampleExpressions :: [(String, Expr)]
exampleExpressions =
  [ ("単純な加算", Add (Lit 10) (Lit 20))
  , ("乗算と除算", Div (Mul (Lit 6) (Lit 7)) (Lit 2))
  , ("状態の取得", Add Get (Lit 5))
  , ("状態の更新", Put (Add Get (Lit 1)))
  , ("ゼロ除算エラー", Div (Lit 10) (Lit 0))
  , ("非決定的選択", Choose (Lit 1) (Lit 2))
  , ("複雑な例", Add
      (Choose (Lit 10) (Lit 20))
      (Put (Add Get (Lit 1)))
    )
  ]

-- テスト実行関数
runExamples :: IO ()
runExamples = do
  let env = []
      initState = 0
  mapM_ (\(name, expr) -> do
    putStrLn $ "--- " ++ name ++ " ---"
    case runWithAllEffects expr env initState of
      Right results ->
        mapM_ (\(value, state) ->
          putStrLn $ "  結果: " ++ show value ++ ", 状態: " ++ show state
        ) results
      Left err ->
        putStrLn $ "  エラー: " ++ err
    ) exampleExpressions

-- Reml との比較メモ:
--
-- 1. **効果の表現**
--    Reml: effect State<S> { operation get() -> S; operation put(s: S) -> () }
--    Haskell: StateT Int (ExceptT String []) a（モナド変換子スタック）
--    - Reml は言語レベルで効果を定義
--    - Haskell は型クラス（Monad）で効果を合成
--
-- 2. **ハンドラーの実装**
--    Reml: handler state_handler<A>(init) for State<S> { ... }
--    Haskell: evalStateT/runExceptT でモナドを順に実行
--    - Reml はハンドラーが明示的で再利用可能
--    - Haskell は型レベルで効果を積み重ねるが、順序変更が困難
--
-- 3. **非決定性の扱い**
--    Reml: choose_handler で分岐を自動収集
--    Haskell: lift . lift でリストモナドに到達（煩雑）
--    - Reml は分岐が自然に追跡される
--    - Haskell は lift を多用する必要がある
--
-- 4. **型推論**
--    Reml: 効果が型レベルで推論される
--    Haskell: 型注釈が必要（特にモナド変換子）
--    - どちらも型安全だが、Haskell は型が複雑
--
-- 5. **可読性**
--    Reml: with State<Int>, Except<String>, Choose で効果が明確
--    Haskell: StateT Int (ExceptT String []) a が冗長
--    - Reml の方が効果の意図が分かりやすい
--
-- 6. **効果の順序変更**
--    Reml: ハンドラーの順序を簡単に入れ替え可能
--    Haskell: モナド変換子の順序を変えると型が変わり、コード全体の書き換えが必要
--    - Reml の方が柔軟性が高い
--
-- **結論**:
-- Haskell のモナド変換子は強力だが、複雑な効果の合成では煩雑になる。
-- Reml の代数的効果システムはより直感的で、効果の順序変更が容易。
-- 特に lift を多用する必要がない点で、Reml の方が記述性に優れる。

-- テスト実行例（main 関数）
-- main :: IO ()
-- main = runExamples